### PR TITLE
Fixing the threading issue in the syslog monitor

### PR DESCRIPTION
**Pull request by alesnovak-s1**
Requests are read in a thread pool first to release the socket fd as soon as possible to ensure higher throughput.
The data is handled in a separate thread pool.

https://sentinelone.atlassian.net/browse/DTIN-3231

### CODE REVIEW

File: scalyr_agent/builtin_monitors/thread_pool.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import threading
```
Comment (tomaz-s1) [inline scalyr_agent/builtin_monitors/thread_pool.py:1]: Let's please add ASF license header to this file.

File: scalyr_agent/builtin_monitors/syslog_monitor.py [hunk 2]
```diff
@@ -18,6 +18,10 @@
 from __future__ import absolute_import
```
Comment (tomaz-s1) [inline scalyr_agent/builtin_monitors/syslog_monitor.py:18]: It would be good to add a docstring to this module to explain to the developers how request processing works and that we use a thread pool for processing requests and reading data from the socket.

File: scalyr_agent/builtin_monitors/syslog_monitor.py [hunk 3]
```diff
@@ -570,77 +576,113 @@ class SyslogUDPHandler(six.moves.socketserver.BaseRequestHandler):
...
     def handle(self):
         data = six.ensure_text(self.request[0].strip(), "utf-8", errors="ignore")
         extra = {
             "proto": "udp",
             "srcip": self.client_address[0],
             "destport": self.server.server_address[1],
         }
-        self.server.syslog_handler.handle(data, extra)
 
+        self.__request_processing_executor.submit(
```
Comment (tomaz-s1) [inline scalyr_agent/builtin_monitors/syslog_monitor.py:602]: Since `handle` is now called inside a thread, do we need any kind of locking around this method?

It appears we already have a lock in https://github.com/scalyr/scalyr-agent-2/blob/888447321333b7f817739cb06806a461188b3ba7/scalyr_agent/builtin_monitors/syslog_monitor.py#L1583 for file creation related reasons and all `handle()` method does is pass data to the logger (https://github.com/scalyr/scalyr-agent-2/blob/888447321333b7f817739cb06806a461188b3ba7/scalyr_agent/builtin_monitors/syslog_monitor.py#L1738).

I believe we already have a lock in our custom logger methods, but it would be good to confirm there are no race conditions - starting a monitor and sending a lot of data to in in parallel manner and verifying that syslog monitor log file is correct aka that data is not interleaved.

File: scalyr_agent/builtin_monitors/thread_pool.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+from concurrent.futures import ThreadPoolExecutor
+
+from scalyr_agent import scalyr_logging
+
+global_log = scalyr_logging.getLogger(__name__)
+
+class ExecutorMixIn:
+    def __init__(self):
+        # Using 4 threads for processing requests, because of GIL and CPU bound tasks higher number would not help process the data faster.
+        self._request_processing_executor = ThreadPoolExecutorFactory.get_singleton("request_processing_executor", max_workers=4)
```
Comment (tomaz-s1) [inline scalyr_agent/builtin_monitors/thread_pool.py:11]: Do we want to make `max_workers` configurable?

Perhaps it's a good idea in case someone hits a race condition or similar so we can provide them with a workaround to set this value to `1`.

File: scalyr_agent/monitors_manager.py [hunk 5]
```diff
@@ -208,6 +210,10 @@ def stop_manager(self, wait_on_join=True, join_timeout=5):
 
         start_time = time.time()
 
+        log.info("Stopping shared thread pools")
+        ThreadPoolExecutorFactory.shutdown(wait=False, cancel_futures=True)
```
Comment (tomaz-s1) [inline scalyr_agent/monitors_manager.py:214]: To me it seems like that this should live on the syslog monitor `stop()` method and not inside the MonitorsManager itself.

Comment (alesnovak-s1) [reply scalyr_agent/monitors_manager.py:214]: The thread pools are shared across the syslog monitors. That's why it needs to be here.

Comment (tomaz-s1) [reply scalyr_agent/monitors_manager.py:214]: Ah, ok, makes sense, thanks for clarifying.

Could this cause any issues since each monitor is running in a separate thread already?

Comment (alesnovak-s1) [reply scalyr_agent/monitors_manager.py:214]: That's a good question.
ThreadPool.submit (this method is thread safe) adds a callable item with parameters to a queue and starts a new thread and adds it to an internal pool if num_threads < self._max_workers.
What I was worried about is creating and starting a long running nested thread - possibly blocking joining of the parent thread. But that's not the case in python.
It would be a better design not to call processing_executor.submit from inside thread from a request_reading_executor pool, but abstractions around BaseServer don't allow it. 
Ideally we should use an async server like tornado to handle these requests, but that's for another PR :)

File: scalyr_agent/builtin_monitors/thread_pool.py (hunk 4 above)
Comment (alesnovak-s1) [reply scalyr_agent/builtin_monitors/thread_pool.py:11]: This was a laziness on my part, I will make it configurable :)

File: scalyr_agent/builtin_monitors/syslog_monitor.py (hunk 3 above)
Comment (tomaz-s1) [reply scalyr_agent/builtin_monitors/syslog_monitor.py:602]: Thanks for the clarification :)

File: scalyr_agent/monitors_manager.py (hunk 5 above)
Comment (alesnovak-s1) [reply scalyr_agent/monitors_manager.py:214]: > To me it seems like that this should live on the syslog monitor `stop()` method and not inside the MonitorsManager itself.

Looking more into the code of syslog monitors, I've realized you're right.
Since SyslogMonitor is in charge of starting and stopping all Syslog Servers, it makes sense for it to be responsible for creation and shutting down of the pools.

Comment (alesnovak-s1) [reply scalyr_agent/monitors_manager.py:214]: In the end it really needs to be in the Monitor Manager, I understood the Syslog Monitor wrong. It starts servers for one monitor configuration only.

**Comment by alesnovak-s1** [comment]
@tomaz-s1 
I have added configuration, changed tests to follow the new api of some of the classes (Syslog Request and Parser).
And now to ensure the request data is processed in the correct order, the worker submitted to the processing executor gets his data from a Queue (one queue and one worker per request). 
Please review again.

**Comment by tomaz-s1** [comment]
@alesnovak-s1 Thanks.

Have we done any end to end testing with agent running multiple monitor instances and producer sending data in parallel (e.g. using https://github.com/scalyr/syslog-collector/tree/main/data-and-load-generator tool)?

File: scalyr_agent/builtin_monitors/syslog_monitor.py [hunk 6]
```diff
@@ -570,77 +583,110 @@ class SyslogUDPHandler(six.moves.socketserver.BaseRequestHandler):
...
-            return None
-        except NON_BLOCKING_SOCKET_DATA_NOT_READY_EXCEPTIONS:
-            return None
+                raise SocketClosed()
+        except (socket.timeout, *NON_BLOCKING_SOCKET_DATA_NOT_READY_EXCEPTIONS) as e:
+            raise SocketNotReadyException(e)
         except socket.error as e:
             if e.errno == errno.EAGAIN:
-                return None
+                raise SocketNotReadyException(e)
```
Comment (tomaz-s1) [inline scalyr_agent/builtin_monitors/syslog_monitor.py:643]: Just curious why this change is needed? In the past, we had bug related to this code path throwing, but this could have changed with the multi threading model.

**Review by tomaz-s1 — APPROVED** [review]
LGTM, but I would feel more comfortable if we performed more end to end testing with multiple monitor instances and highly parallel producer.

File: scalyr_agent/builtin_monitors/syslog_monitor.py (hunk 6 above)
Comment (alesnovak-s1) [reply scalyr_agent/builtin_monitors/syslog_monitor.py:643]: Before the method returned None for both cases - when data on socket were not ready and when the socket was closed.
And you had to guess what happened. 
The exceptions are path of the code flow and are expected and catched right away, avoiding any propagating.

Comment (alesnovak-s1) [reply scalyr_agent/builtin_monitors/syslog_monitor.py:643]: And I don't like returning None when something unexpected happened. It's quite error-prone.

**Comment by alesnovak-s1** [comment]
> @alesnovak-s1 Thanks.
> 
> Have we done any end to end testing with agent running multiple monitor instances and producer sending data in parallel (e.g. using https://github.com/scalyr/syslog-collector/tree/main/data-and-load-generator tool)?

Yes, I have a config with multiple Syslog (~15) monitors configured with both udp and tcp ports.
I was using loggen and syslog-loadgen and sending tons of data to all the ports in parallel, several processes per port.
